{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "\n",
    "# PheKnowLator - Mapping Data\n",
    "\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "**Author:** [TJCallahan](https://mail.google.com/mail/u/0/?view=cm&fs=1&tf=1&to=callahantiff@gmail.com)  \n",
    "**GitHub Repository:** [PheKnowLator](https://github.com/callahantiff/PheKnowLator/wiki)  \n",
    "\n",
    "**Purpose:** This notebook serves as a file to generate mapping and filtering data for the PheKnowLator project. The script creates each of the mapping and/or filtering data sources described on the [Data Sources](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-and-filtering-data) Wiki page. \n",
    "\n",
    "**Assumptions:** The script assumes that there are files, which need further processing and are located in the `./resources/data_maps/unprocessed_data/` directory.\n",
    "\n",
    "\n",
    "***\n",
    "\n",
    "### Table of Contents\n",
    "* [MESH-ChEBI](#mesh-chebi)  \n",
    "* [Ensembl Gene - Ensembl Transcript](#ensemblgene-ensembltranscript)  \n",
    "* [Ensembl Gene - ENTREZ Gene](#ensemblgene-entrezgene)  \n",
    "* [Ensembl Gene - Uniprot Accession](#ensemblgene-uniprot)  \n",
    "* [Ensembl Protein - Uniprot Accession](#ensemblprotein-uniprot)  \n",
    "* [HPA Tissue/Cells - UBERON + Cell Ontology](#hpa-uberon)  \n",
    "* [Uniprot Accession - Protein Ontology](#uniprot-pro)  \n",
    "* [Ensembl Protein - Protein Ontology](#ensemblprotein-pro)  \n",
    "\n",
    "***\n",
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import needed libraries\n",
    "import glob\n",
    "import pandas\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create Helper Function\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_processor(filepath: str, row_splitter: str, column_list: list, output_name: str, line_splitter: str = ''):\n",
    "    \"\"\"Reads in a file using input file path and reduces the file to only include specific columns specified by the\n",
    "    input var. The reduced file is then saved as a text file and written to the `/resources/data_maps' directory.\n",
    "\n",
    "    Args:\n",
    "        filepath (str): A string that points to the location of a temp mapping file that needs to be processed.\n",
    "        row_splitter (str): A string that contains a character used to split rows.\n",
    "        column_list (list): A list that contains two numbers, which correspond to indices in the input data file and\n",
    "                            which appear in the order of write preference.\n",
    "        output_name (str): A string naming the processed data file.\n",
    "        line_splitter (str): A character used to separate multiple data points from a string. Defaults to an empty\n",
    "                             string which is used to indicate the string contains a single value.\n",
    "\n",
    "    Return:\n",
    "        None.\n",
    "    \"\"\"\n",
    "\n",
    "    # read in data\n",
    "    data = open(filepath).readlines()\n",
    "\n",
    "    # process and write out data\n",
    "    with open('../../resources/data_maps/{filename}.txt'.format(filename=output_name), 'w') as outfile:\n",
    "\n",
    "        for line in data:\n",
    "            subj = line.split(row_splitter)[column_list[0]]\n",
    "            obj = line.split(row_splitter)[column_list[1]]\n",
    "\n",
    "            if subj != '' and obj != '':\n",
    "                for i in [subj.split(line_splitter) if line_splitter != '' else [subj]][0]:\n",
    "                    for j in [obj.split(line_splitter) if line_splitter != '' else [obj]][0]:\n",
    "                        outfile.write(i.strip() + '\\t' + j.strip() + '\\n')\n",
    "\n",
    "    outfile.close()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "\n",
    "### MESH - ChEBI <a class=\"anchor\" id=\"mesh-chebi\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-mesh-to-chebi](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-mesh-to-chebi)  \n",
    "\n",
    "**Purpose:** This script assumes that the `NCBO_rest_api.py` script was run and the data generated from this file was written to `./resources/data_maps/temp`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../../resources/data_maps/MESH_CHEBI_MAP.txt', 'w') as out:\n",
    "    for filename in glob.glob('../../resources/data_maps/temp/*.txt'):\n",
    "        for row in list(filter(None, open(filename, 'r').read().split('\\n'))):\n",
    "            mesh = '_'.join(row.split('\\t')[0].split('/')[-2:])\n",
    "            chebi = row.split('\\t')[1].split('/')[-1]\n",
    "            out.write(mesh + '\\t' + chebi + '\\n')\n",
    "\n",
    "out.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MESH_C535085</td>\n",
       "      <td>CHEBI_133814</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MESH_C008574</td>\n",
       "      <td>CHEBI_17221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MESH_C492482</td>\n",
       "      <td>CHEBI_34581</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MESH_C007556</td>\n",
       "      <td>CHEBI_135978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MESH_C500395</td>\n",
       "      <td>CHEBI_29138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>MESH_C028026</td>\n",
       "      <td>CHEBI_15439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>MESH_C560044</td>\n",
       "      <td>CHEBI_68089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>MESH_C035745</td>\n",
       "      <td>CHEBI_138829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>MESH_C511148</td>\n",
       "      <td>CHEBI_6566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MESH_C050445</td>\n",
       "      <td>CHEBI_28160</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0             1\n",
       "0  MESH_C535085  CHEBI_133814\n",
       "1  MESH_C008574   CHEBI_17221\n",
       "2  MESH_C492482   CHEBI_34581\n",
       "3  MESH_C007556  CHEBI_135978\n",
       "4  MESH_C500395   CHEBI_29138\n",
       "5  MESH_C028026   CHEBI_15439\n",
       "6  MESH_C560044   CHEBI_68089\n",
       "7  MESH_C035745  CHEBI_138829\n",
       "8  MESH_C511148    CHEBI_6566\n",
       "9  MESH_C050445   CHEBI_28160"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/MESH_CHEBI_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### Ensembl Gene - Ensembl Transcript <a class=\"anchor\" id=\"ensemblgene-ensembltranscript\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-transcript-protein-and-gene-identifiers](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-transcript-protein-and-gene-identifiers)  \n",
    "\n",
    "**Purpose:** This script assumes that the `HUMAN_9606_idmapping_selected.tab.gz` file from the [Uniprot Knolwedge Base](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping_selected.tab.gz) was downloaded and saved to the `./resources/data_maps/unprocessed_data/` directory.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_processor(filepath='../../resources/data_maps/unprocessed_data/HUMAN_9606_idmapping_selected.tab',\n",
    "                   row_splitter='\\t',\n",
    "                   column_list=[18, 19],\n",
    "                   output_name='ENSEMBL_GENE_ENSEMBL_TRANSCRIPT_MAP',\n",
    "                   line_splitter=';')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000166913</td>\n",
       "      <td>ENST00000353703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000166913</td>\n",
       "      <td>ENST00000372839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>ENST00000264335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>ENST00000571732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>ENST00000616643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>ENST00000627231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000274474</td>\n",
       "      <td>ENST00000264335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000274474</td>\n",
       "      <td>ENST00000571732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000274474</td>\n",
       "      <td>ENST00000616643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000274474</td>\n",
       "      <td>ENST00000627231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0                1\n",
       "0  ENSG00000166913  ENST00000353703\n",
       "1  ENSG00000166913  ENST00000372839\n",
       "2  ENSG00000108953  ENST00000264335\n",
       "3  ENSG00000108953  ENST00000571732\n",
       "4  ENSG00000108953  ENST00000616643\n",
       "5  ENSG00000108953  ENST00000627231\n",
       "6  ENSG00000274474  ENST00000264335\n",
       "7  ENSG00000274474  ENST00000571732\n",
       "8  ENSG00000274474  ENST00000616643\n",
       "9  ENSG00000274474  ENST00000627231"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/ENSEMBL_GENE_ENSEMBL_TRANSCRIPT_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### Ensembl Gene - ENTREZ Gene <a class=\"anchor\" id=\"ensemblgene-entrezgene\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-transcript-protein-and-gene-identifiers](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-transcript-protein-and-gene-identifiers)  \n",
    "\n",
    "**Purpose:** This script assumes that the `HUMAN_9606_idmapping_selected.tab.gz` file from the [Uniprot Knolwedge Base](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping_selected.tab.gz) was downloaded and saved to the `./resources/data_maps/unprocessed_data/` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_processor(filepath='../../resources/data_maps/unprocessed_data/HUMAN_9606_idmapping_selected.tab',\n",
    "                   row_splitter='\\t',\n",
    "                   column_list=[18, 2],\n",
    "                   output_name='ENSEMBL_GENE_ENTREZ_GENE_MAP',\n",
    "                   line_splitter=';')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000166913</td>\n",
       "      <td>7529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>7531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000274474</td>\n",
       "      <td>7531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000128245</td>\n",
       "      <td>7533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000170027</td>\n",
       "      <td>7532</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000175793</td>\n",
       "      <td>2810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000134308</td>\n",
       "      <td>10971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000164924</td>\n",
       "      <td>7534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000110455</td>\n",
       "      <td>84680</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000205126</td>\n",
       "      <td>390110</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0       1\n",
       "0  ENSG00000166913    7529\n",
       "1  ENSG00000108953    7531\n",
       "2  ENSG00000274474    7531\n",
       "3  ENSG00000128245    7533\n",
       "4  ENSG00000170027    7532\n",
       "5  ENSG00000175793    2810\n",
       "6  ENSG00000134308   10971\n",
       "7  ENSG00000164924    7534\n",
       "8  ENSG00000110455   84680\n",
       "9  ENSG00000205126  390110"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/ENSEMBL_GENE_ENTREZ_GENE_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### Ensembl Gene - Uniprot Accession <a class=\"anchor\" id=\"ensemblgene-uniprot\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-transcript-protein-and-gene-identifiers](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-transcript-protein-and-gene-identifiers)  \n",
    "\n",
    "**Purpose:** This script assumes that the `HUMAN_9606_idmapping_selected.tab.gz` file from the [Uniprot Knolwedge Base](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping_selected.tab.gz) was downloaded and saved to the `./resources/data_maps/unprocessed_data/` directory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_processor(filepath='../../resources/data_maps/unprocessed_data/HUMAN_9606_idmapping_selected.tab',\n",
    "                   row_splitter='\\t',\n",
    "                   column_list=[18, 0],\n",
    "                   output_name='ENSEMBL_GENE_UNIPROT_ACCESSION_MAP',\n",
    "                   line_splitter=';')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000166913</td>\n",
       "      <td>P31946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000108953</td>\n",
       "      <td>P62258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000274474</td>\n",
       "      <td>P62258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000128245</td>\n",
       "      <td>Q04917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000170027</td>\n",
       "      <td>P61981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000175793</td>\n",
       "      <td>P31947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000134308</td>\n",
       "      <td>P27348</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000164924</td>\n",
       "      <td>P63104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000110455</td>\n",
       "      <td>Q96QU6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000205126</td>\n",
       "      <td>Q4AC99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0       1\n",
       "0  ENSG00000166913  P31946\n",
       "1  ENSG00000108953  P62258\n",
       "2  ENSG00000274474  P62258\n",
       "3  ENSG00000128245  Q04917\n",
       "4  ENSG00000170027  P61981\n",
       "5  ENSG00000175793  P31947\n",
       "6  ENSG00000134308  P27348\n",
       "7  ENSG00000164924  P63104\n",
       "8  ENSG00000110455  Q96QU6\n",
       "9  ENSG00000205126  Q4AC99"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/ENSEMBL_GENE_UNIPROT_ACCESSION_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### Ensembl Protein - Uniprot Accession <a class=\"anchor\" id=\"ensemblprotein-uniprot\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-transcript-protein-and-gene-identifiers](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-transcript-protein-and-gene-identifiers)  \n",
    "\n",
    "**Purpose:** This script assumes that the `HUMAN_9606_idmapping_selected.tab.gz` file from the [Uniprot Knolwedge Base](ftp://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping/by_organism/HUMAN_9606_idmapping_selected.tab.gz) was downloaded and saved to the `./resources/data_maps/unprocessed_data/` directory.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_processor(filepath='../../resources/data_maps/unprocessed_data/HUMAN_9606_idmapping_selected.tab',\n",
    "                   row_splitter='\\t',\n",
    "                   column_list=[20, 0],\n",
    "                   output_name='ENSEMBL_PROTEIN_UNIPROT_ACCESSION_MAP',\n",
    "                   line_splitter=';')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSP00000300161</td>\n",
       "      <td>P31946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSP00000361930</td>\n",
       "      <td>P31946</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSP00000264335</td>\n",
       "      <td>P62258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSP00000461762</td>\n",
       "      <td>P62258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSP00000481059</td>\n",
       "      <td>P62258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSP00000487356</td>\n",
       "      <td>P62258</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSP00000248975</td>\n",
       "      <td>Q04917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSP00000306330</td>\n",
       "      <td>P61981</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSP00000340989</td>\n",
       "      <td>P31947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSP00000238081</td>\n",
       "      <td>P27348</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0       1\n",
       "0  ENSP00000300161  P31946\n",
       "1  ENSP00000361930  P31946\n",
       "2  ENSP00000264335  P62258\n",
       "3  ENSP00000461762  P62258\n",
       "4  ENSP00000481059  P62258\n",
       "5  ENSP00000487356  P62258\n",
       "6  ENSP00000248975  Q04917\n",
       "7  ENSP00000306330  P61981\n",
       "8  ENSP00000340989  P31947\n",
       "9  ENSP00000238081  P27348"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/ENSEMBL_PROTEIN_UNIPROT_ACCESSION_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### HPA Tissue/Cells - UBERON + Cell Ontology <a class=\"anchor\" id=\"hpa-uberon\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-human-protein-atlas-tissues-and-cell-types](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-human-protein-atlas-tissues-and-cell-types)  \n",
    "\n",
    "**Purpose:** This script assumes that the `rna_tissue_consensus.tsv` and `normal_tissue.tsv` files from the [Human Protein Atlas](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#human-protein-atlas) were downloaded and saved to the `./resources/data_maps/unprocessed_data/` directory.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# abnormal tissue\n",
    "abnormal_tissue = []\n",
    "for line in open('../../resources/data_maps/unprocessed_data/rna_tissue_consensus.tsv').readlines():\n",
    "    abnormal_tissue.append(line.split('\\t')[2].strip())\n",
    "\n",
    "# normal tissue\n",
    "normal_tissue = []\n",
    "\n",
    "for line in open('../../resources/data_maps/unprocessed_data/normal_tissue.tsv').readlines():\n",
    "    normal_tissue.append(line.split('\\t')[2].strip() + ' - ' + line.split('\\t')[3].strip())\n",
    "\n",
    "# combine normal and abnormal tissue and cells into single list\n",
    "combo = set(abnormal_tissue + normal_tissue)\n",
    "\n",
    "# write results\n",
    "with open('../../resources/data_maps/unprocessed_data/HPA_tissues.txt', 'w') as outfile:\n",
    "    for x in combo:\n",
    "        outfile.write(x.strip() + '\\n')\n",
    "\n",
    "outfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TISSUE</th>\n",
       "      <th>CELL TYPE</th>\n",
       "      <th>ONTOLOGY</th>\n",
       "      <th>ONTOLOGY ID</th>\n",
       "      <th>ONTOLOGY LABEL</th>\n",
       "      <th>MAPPING</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>None</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001013</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>ZOOMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>adipocytes</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001013</td>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>ZOOMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>adipocytes</td>\n",
       "      <td>CL</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0001070</td>\n",
       "      <td>fat cell</td>\n",
       "      <td>Manual</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>None</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002369</td>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>ZOOMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>cells in zona fasciculata</td>\n",
       "      <td>UBERON</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002054</td>\n",
       "      <td>zona fasciculata of adrenal gland</td>\n",
       "      <td>ZOOMA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            TISSUE                   CELL TYPE ONTOLOGY  \\\n",
       "0   adipose tissue                        None   UBERON   \n",
       "1  adipose tissue                   adipocytes   UBERON   \n",
       "2  adipose tissue                   adipocytes       CL   \n",
       "3    adrenal gland                        None   UBERON   \n",
       "4   adrenal gland    cells in zona fasciculata   UBERON   \n",
       "\n",
       "                                     ONTOLOGY ID  \\\n",
       "0  http://purl.obolibrary.org/obo/UBERON_0001013   \n",
       "1  http://purl.obolibrary.org/obo/UBERON_0001013   \n",
       "2      http://purl.obolibrary.org/obo/CL_0001070   \n",
       "3  http://purl.obolibrary.org/obo/UBERON_0002369   \n",
       "4  http://purl.obolibrary.org/obo/UBERON_0002054   \n",
       "\n",
       "                      ONTOLOGY LABEL MAPPING  \n",
       "0                     adipose tissue   ZOOMA  \n",
       "1                     adipose tissue   ZOOMA  \n",
       "2                           fat cell  Manual  \n",
       "3                      adrenal gland   ZOOMA  \n",
       "4  zona fasciculata of adrenal gland   ZOOMA  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read back in mapped tissue/cell data\n",
    "data = pandas.read_excel(open('../../resources/data_maps/unprocessed_data/zooma_tissue_cell_mapping_04DEC2019.xlsx',\n",
    "                              'rb'),\n",
    "                         sheet_name='zooma_tissue_cell_mapping_04DEC',\n",
    "                         header=0)\n",
    "\n",
    "data.fillna('None', inplace=True)\n",
    "\n",
    "# prerview data\n",
    "data.head(n=5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat data and write it out\n",
    "with open('../../resources/data_maps/HPA_TISSUE_CELL_MAP.txt', 'w') as outfile:\n",
    "    for idx, row in data.iterrows():\n",
    "\n",
    "        if row['TISSUE'] != 'None':\n",
    "            outfile.write(str(row['TISSUE']).strip() + '\\t' + str(row['ONTOLOGY ID']).strip() + '\\n')\n",
    "\n",
    "        if row['CELL TYPE'] != 'None':\n",
    "            outfile.write(str(row['CELL TYPE']).strip() + '\\t' + str(row['ONTOLOGY ID']).strip() + '\\n')\n",
    "\n",
    "outfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>adipocytes</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0001013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>adipose tissue</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0001070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>adipocytes</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0001070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>cells in zona fasciculata</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>adrenal gland</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0002136</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>cells in zona fasciculata</td>\n",
       "      <td>http://purl.obolibrary.org/obo/CL_0002136</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0                                              1\n",
       "0             adipose tissue  http://purl.obolibrary.org/obo/UBERON_0001013\n",
       "1             adipose tissue  http://purl.obolibrary.org/obo/UBERON_0001013\n",
       "2                 adipocytes  http://purl.obolibrary.org/obo/UBERON_0001013\n",
       "3             adipose tissue      http://purl.obolibrary.org/obo/CL_0001070\n",
       "4                 adipocytes      http://purl.obolibrary.org/obo/CL_0001070\n",
       "5              adrenal gland  http://purl.obolibrary.org/obo/UBERON_0002369\n",
       "6              adrenal gland  http://purl.obolibrary.org/obo/UBERON_0002054\n",
       "7  cells in zona fasciculata  http://purl.obolibrary.org/obo/UBERON_0002054\n",
       "8              adrenal gland      http://purl.obolibrary.org/obo/CL_0002136\n",
       "9  cells in zona fasciculata      http://purl.obolibrary.org/obo/CL_0002136"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/HPA_TISSUE_CELL_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### Uniprot Accession - Protein Ontology <a class=\"anchor\" id=\"uniprot-pro\"></a>\n",
    "\n",
    "**Wiki Page:** [mapping-protein-identifiers](https://github.com/callahantiff/PheKnowLator/wiki/v2-Data-Sources/#mapping-protein-identifiers)  \n",
    "\n",
    "**Purpose:** This script assumes that the `promapping.txt` file from the [Pro Consortium](https://proconsortium.org/download/current/) was downloaded and saved to the `./resources/data_maps/unprocessed_data/` directory.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = open('../../resources/data_maps/unprocessed_data/promapping.txt').readlines()\n",
    "\n",
    "# reformat data and write it out\n",
    "with open('../../resources/data_maps/UNIPROT_ACCESSION_PRO_MAP.txt', 'w') as outfile:\n",
    "    for line in data:\n",
    "        row = line.split('\\t')\n",
    "\n",
    "        if row[1].startswith('UniProtKB'):\n",
    "            outfile.write(row[0].strip().replace(':', '_') + '\\t' + row[1].strip().split(':')[-1] + '\\n')\n",
    "\n",
    "outfile.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PR_000000005</td>\n",
       "      <td>P37173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PR_000000005</td>\n",
       "      <td>P38438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PR_000000005</td>\n",
       "      <td>Q62312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PR_000000005</td>\n",
       "      <td>Q90999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PR_000000007</td>\n",
       "      <td>F1R709</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PR_000000009</td>\n",
       "      <td>Q16671</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>PR_000000009</td>\n",
       "      <td>Q62893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PR_000000009</td>\n",
       "      <td>Q8K592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>PR_000000010</td>\n",
       "      <td>O57472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>PR_000000010</td>\n",
       "      <td>Q24025</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0       1\n",
       "0  PR_000000005  P37173\n",
       "1  PR_000000005  P38438\n",
       "2  PR_000000005  Q62312\n",
       "3  PR_000000005  Q90999\n",
       "4  PR_000000007  F1R709\n",
       "5  PR_000000009  Q16671\n",
       "6  PR_000000009  Q62893\n",
       "7  PR_000000009  Q8K592\n",
       "8  PR_000000010  O57472\n",
       "9  PR_000000010  Q24025"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/UNIPROT_ACCESSION_PRO_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<br>\n",
    "\n",
    "***\n",
    "***\n",
    "\n",
    "### Ensembl Protein - Protein Ontology <a class=\"anchor\" id=\"ensemblprotein-pro\"></a>\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pro_kb = open('./resources/data_maps/UNIPROT_ACCESSION_PRO_MAP.txt').readlines()\n",
    "\n",
    "# convert to dictionary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# preview data\n",
    "data = pandas.read_csv('../../resources/data_maps/UNIPROT_ACCESSION_PRO_MAP.txt',\n",
    "                       header = None,\n",
    "                       delimiter = '\\t')\n",
    "\n",
    "data.head(n=10)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.2"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
